Automattic/jetpack
# Backup: Add namespace versioning to Helper_Script_Manager and other classes

Add class versioning using PHP namespaces, to make sure that the plugin always loads the expected version of a class.

## Proposed changes:


When two or more plugins are installed that contain a class with the same name (e.g. the main Jetpack plugin and the Jetpack VaultPress Backup plugin both ship `Automattic\Jetpack\Backup\Helper_Script_Manager` class), the Jetpack autoloader https://github.com/Automattic/jetpack/pull/34297#discussion_r1424048710, and under certain circumstances it might even decide to load an older version of the class.

Specifically, we had issues with the Jetpack VaultPress Backup plugin loading an older version of the `Helper_Script_Manager` class using the newer code, in which case an `abspath` property in `Helper_Script_Manager::install_helper_script()` (https://github.com/Automattic/jetpack/pull/34297) wasn't always returned.

* p1703028095475669-slack-CS8UYNPEE

To avoid that behavior, and always load the *exact same* version of the class that has been shipped with the plugin, we're adding simple namespace-based versioning to `Helper_Script_Manager`, backup plugin, and other related code.

All of the classes that make up the package/plugin got moved to either the `Automattic\Jetpack\Backup\V0001` or the `Automattic\Jetpack\Transport_Helper\V0001` namespace. Code that uses said classes refer to them using the versioned namespace (e.g. `Automattic\Jetpack\Backup\V0001\Helper_Script_Manager::install_helper_script()`, so that once the plugin gets installed, it always loads and uses a specific version of its dependencies (and not an older/newer one, which it might be incompatible with).

Older versions of the Jetpack plugins don't use these namespaces (e.g. they refer to the `Helper_Script_Manager` class as `Automattic\Jetpack\Backup\Helper_Script_Manager`), so if the plugin with versioned namespaces gets installed together with an older plugin without those, both should work fine.

If we ever need to change the plugin, the process is somewhat cumbersome yet straightforward:

1. Do the changes that you need;
2. Search for `V0001`, replace it with `V0002` everywhere.

to make sure that older versions of the plugin load and use the `V0001` version of the relevant code, and the newer versions use `V0002`.

Relevant discussions:

* peaFOp-2ar-p2
* https://github.com/Automattic/jetpack/pull/34297#discussion_r1424048710

### Other information:


## Jetpack product discussion



peaFOp-2ar-p2

## Does this pull request change what data or activity we track or use?



No.

## Testing instructions:






Just like with https://github.com/Automattic/jetpack/pull/34297, run a Jetpack VaultPress Backup backup without credentials set up, so that the tool is then forced to upload the helper script via Jetpack API and thus trigger this code.

Make sure that the `abspath` property gets always returned in the `install_helper_script()` results.

Also, test with multiple versions/forms of various Jetpack plugins installed, e.g. the newer Jetpack VaultPress Backup plugin (with this patch), and the older main Jetpack plugin (without this patch); make sure that you still always get the `abspath` back.
